### DISCUSSION TITLE
Why is asset precompile failing for me with rails 7.1 alpha?

### DISCUSSION THREAD

**Discussion by floydj** [basecamp/kamal#341] [Q&A]
Trying out mrsk with a very basic 7.1 app. Have changed little in the default deploy.yml app, just my username for Docker hub and my server IP.

All seemed to be going well, but `mrsk deploy` fails with the following:

```
#22 [linux/amd64 build 9/9] RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile
#22 1.354 yarn install v1.22.10
#22 1.487 [1/4] Resolving packages...
#22 1.616 success Already up-to-date.
#22 1.622 Done in 0.28s.
#22 1.864 yarn run v1.22.10
#22 1.955 $ esbuild app/javascript/*.* --bundle --sourcemap --outdir=app/assets/builds --public-path=/assets
#22 1.983 /usr/bin/env: ‘node’: Text file busy
#22 1.997 error Command failed with exit code 126.
#22 1.998 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
#22 2.035 bin/rails aborted!
#22 2.035 jsbundling-rails: Command build failed, ensure `yarn build` runs without errors
#22 2.036
#22 2.036 Tasks: TOP => assets:precompile => javascript:build
#22 2.037 (See full trace by running task with --trace)
#22 ERROR: process "/bin/sh -c SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile" did not complete successfully: exit code: 1

Dockerfile:51
--------------------
  49 |
  50 |     # Precompiling assets for production without requiring secret RAILS_MASTER_KEY
  51 | >>> RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile
  52 |
  53 |
--------------------
ERROR: failed to solve: process "/bin/sh -c SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile" did not complete successfully: exit code: 1
```

Any ideas on how I can troubleshoot further? I've logged in direct with SSH and verified that yarn is installed. I couldn't figure out how to try to run line 51 direct on the server.

**Comment by intrip** [comment]
Looks like a `yarn build` failure. What's the full output if you add `--trace` as suggested?

**Reply by floydj** [reply]
The trace doesn't really add much that's helpful, at least to me. 

```
#22 [linux/amd64 build 9/9] RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile --trace
#22 0.917 ** Invoke assets:precompile (first_time)
#22 0.917 ** Invoke assets:environment (first_time)
#22 0.918 ** Execute assets:environment
#22 0.918 ** Invoke environment (first_time)
#22 0.918 ** Execute environment
#22 1.089 ** Invoke javascript:build (first_time)
#22 1.089 ** Invoke javascript:install (first_time)
#22 1.089 ** Execute javascript:install
#22 1.315 yarn install v1.22.10
#22 1.477 [1/4] Resolving packages...
#22 1.658 success Already up-to-date.
#22 1.675 Done in 0.37s.
#22 1.734 ** Execute javascript:build
#22 2.053 yarn run v1.22.10
#22 2.165 $ esbuild app/javascript/*.* --bundle --sourcemap --outdir=app/assets/builds --public-path=/assets
#22 2.212 /usr/bin/env: ‘node’: Text file busy
#22 2.248 error Command failed with exit code 126.
#22 2.258 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
#22 2.305 bin/rails aborted!
#22 2.305 jsbundling-rails: Command build failed, ensure `yarn build` runs without errors
```

**Reply by phillmv** [reply]
What happens when you run the esbuild step from after sshing into the docker? `text file busy` would seem to indicate that something is overwriting `node`, which is a strange error.

**Reply by floydj** [reply]
How can I ssh into the docker container? Once the mrsk task finishes, no docker containers are listed on the server.

**Reply by phillmv** [reply]
Sorry, I picked the wrong words.

What happens when you create a build stage (`FROM foo as just-before-esbuild`) immediately before the esbuild compilation, build the docker image to that target `docker build --target just-before-esbuild -t namehere:latest` then interactively start bash in it (`docker run -it namehere:latest bash`) and THEN try running the esbuild step from there?

it should give you more debugging options rather than just watching the build fail.

**Reply by floydj** [reply]
Sorry for the delay.... How do I do that in the context of MRSK? I'm not sure how to "interactively start bash" with MRSK.

**Comment by scart88** [comment]
You can't ssh into a docker container.
You will need to shh into your linux servers and run `docker ps` to see which containers are running, however, from your error, it looks doesn't look like you are on that step yet.

You are having issues with building the docker image.
After the docker image is uploaded on your registry (docker hub, ghrc, gitlab...), mrsk will pull it to your server and will start the process.

I had a very good experience with https://github.com/fly-apps/dockerfile-rails , so you can try it to build your docker image before using it with the mrsk.

**Reply by floydj** [reply]
Right. It's failing on my asset compilation. I'll take a look at the repo you linked. Thank you

**Comment by glaucocustodio** [comment]
I am getting the same error with `rails 7.1.2`

**Reply by stmpjmpr** [reply]
A similar error happens with Rails 7.1.2 and Ruby 3.3.0 on Apple Silicon...maybe that's what you're hitting. You can either: build only for your target (assuming that's not Apple Silicon); wait for Ruby 3.3.1, where a fix is scheduled; or apply the patch to the bug yourself and build a custom Ruby. https://bugs.ruby-lang.org/issues/20085

**Reply by glaucocustodio** [reply]
>Rails 7.1.2 and Ruby 3.3.0 on Apple Silicon

that's exactly my case, thanks for pointing the issue
